openadapt/vision.py

```diff
@@ -127,6 +127,34 @@ def refine_masks(masks: list[np.ndarray]) -> list[np.ndarray]:
     logger.info(f"{len(refined_masks)=}")
     return refined_masks
 
+@cache.cache()
+def extract_difference_image(
+    new_image: Image.Image,
+    old_image: Image.Image,
+    tolerance: float = 0.05,
+) -> Image.Image:
+    """Extract the portion of the new image that is different from the old image."""
+    new_image_np = np.array(new_image.convert('L'))
+    old_image_np = np.array(old_image.convert('L'))
+
+    # Compute the SSIM between the two images
+    score, diff = ssim(new_image_np, old_image_np, full=True)
+    diff = (diff * 255).astype("uint8")
+
+    # Threshold the difference image to get the regions that are different
+    thresh = cv2.threshold(diff, 255 * (1 - tolerance), 255, cv2.THRESH_BINARY_INV)[1]
+
+    # Find contours of the different regions
+    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+
+    # Create a mask of the differences
+    mask = np.zeros_like(new_image_np)
+    cv2.drawContours(mask, contours, -1, (255), thickness=cv2.FILLED)
```

Comment:
I believe contours should not be necessary. A more direct approach involves getting the indexes of the pixels whose difference is above a threshold, and setting only those pixels directly.